File: 322-coin-change/coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, 1e9);
        dp[0]=0;
        for(int i=0;i<amount;++i){
            if(dp[i]==1e9) continue;
            for(int c:coins){
                if(i<=amount-c)dp[i+c]=min(dp[i+c],1+dp[i]);
            }
        }
        return dp[amount]!=1e9?dp[amount]:-1;
    }
};

File: 322-coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp= new int[amount+1];
        for(int i=1;i<=amount;i++) dp[i]= amount+1;
        for(int i=1;i<=amount;i++){
            for(int c: coins){
                if(i>=c && dp[i-c]<=amount){
                    dp[i]= Math.min(dp[i],1+dp[i-c]);
                }
            }
        }
        return dp[amount]<=amount?dp[amount]:-1;
    }
}